examples/datadog/main.tf

```diff
@@ -0,0 +1,197 @@
+    # Change this according to your security requirements if needed
+    ingress_cluster_to_node_all_traffic = {
+      description                   = "Cluster API to Nodegroup all traffic"
+      protocol                      = "-1"
+      from_port                     = 0
+      to_port                       = 0
+      type                          = "ingress"
+      source_cluster_security_group = true
+    }
+  }
+
+  managed_node_groups = {
+    mg_5 = {
+      node_group_name      = "managed-ondemand"
+      instance_types       = ["m5.large"]
+      subnet_ids           = module.vpc.private_subnets
+      force_update_version = true
+    }
+  }
+
+  tags = local.tags
+}
+
+module "eks_blueprints_kubernetes_addons" {
+  source = "../../modules/kubernetes-addons"
+
+  eks_cluster_id               = module.eks_blueprints.eks_cluster_id
+  eks_cluster_endpoint         = module.eks_blueprints.eks_cluster_endpoint
+  eks_oidc_provider            = module.eks_blueprints.oidc_provider
+  eks_cluster_version          = module.eks_blueprints.eks_cluster_version
+  eks_worker_security_group_id = module.eks_blueprints.worker_node_security_group_id
+
+  enable_datadog_operator = true
+
+  tags = local.tags
+}
+
+#---------------------------------------------------------------
+# Supporting Resources
+#---------------------------------------------------------------
+
+module "vpc" {
+  source  = "terraform-aws-modules/vpc/aws"
+  version = "~> 3.0"
+
+  name = local.name
+  cidr = local.vpc_cidr
+
+  azs             = local.azs
+  public_subnets  = [for k, v in local.azs : cidrsubnet(local.vpc_cidr, 8, k)]
+  private_subnets = [for k, v in local.azs : cidrsubnet(local.vpc_cidr, 8, k + 10)]
+
+  enable_nat_gateway   = true
+  single_nat_gateway   = true
+  enable_dns_hostnames = true
+
+  # Manage so we can name
+  manage_default_network_acl    = true
+  default_network_acl_tags      = { Name = "${local.name}-default" }
+  manage_default_route_table    = true
+  default_route_table_tags      = { Name = "${local.name}-default" }
+  manage_default_security_group = true
+  default_security_group_tags   = { Name = "${local.name}-default" }
+
+  public_subnet_tags = {
+    "kubernetes.io/cluster/${local.name}" = "shared"
+    "kubernetes.io/role/elb"              = 1
+  }
+
+  private_subnet_tags = {
+    "kubernetes.io/cluster/${local.name}" = "shared"
+    "kubernetes.io/role/internal-elb"     = 1
+  }
+
+  tags = local.tags
+}
+
+#---------------------------------------------------------------
+# Datadog agent
+#---------------------------------------------------------------
+
+resource "kubernetes_secret_v1" "datadog_api_key" {
+  metadata {
+    name      = "datadog-secret"
+    namespace = local.namespace
+  }
+  data = {
+    # This will reveal a secret in the Terraform state
+    api-key = var.datadog_api_key
+  }
+}
+
+resource "kubectl_manifest" "datadog_agent" {
+  yaml_body  = <<YAML
+apiVersion: datadoghq.com/v1alpha1
```

Comment:
convert this to a map so that we can do something like:

```hcl
locals {
	datadog_agent = merge(
		  {
	  		apiVersion = "datadoghq.com/v1alpha1"
			kind = "DatadogAgent"
			metadata:
				name = "datadog"
				namespace = ${local.namespace}
			spec =
				clusterName = ${module.eks_blueprints.eks_cluster_id}
				site = "datadoghq.com"
			....
		  },
		var.helm_config.datadog_agent
	)
}
```

Where users can pass in something like

```hcl
	helm_config = {
		datadog_agent = {
			spec = {
				site = "datadoghq.eu"
			}
		}
	}
```

Reply:
actually, if we move this to a yaml file, we should be able to read it in as a map and then merge it. we should only add the bare minimum so that users are able to extend to suit

```hcl
locals = {
	datadog_agent = merge(yamldecode(templatefile("${path.module}/datadog_agent.yaml", {
			  cluster_name = module.eks_blueprints.eks_cluster_id
			  namespace   = local.namespace
		 })),
		var.helm_config.datadog_agent
	)
}
```

Reply:
I refactored this, as suggested, you may have seen. When I tried to pass in a spec element, as above it overwrites the entire spec so elements such as the `clusterName` and `credentials` are lost if they are not explicitly by the user. Terraform doesn't support deep merging of maps like this. So the `spec` key is simply replaced.

Reply:
@bryantbiggs @askulkarni2 @vara-bonthu @Zvikan  any further thoughts on this refactor and inclusion of the Datadog agent creation?